### ISSUE TITLE
Pie chart supply defaults generates empty marker.colors array

### ISSUE THREAD

**Issue by bpostlethwaite** [plotly/plotly.js#2205]
Offending (?) line of code https://github.com/plotly/plotly.js/blob/599dff93f9923e7ef4d6446de734cfb4eb39d663/src/traces/pie/defaults.js#L38

This results in `marker.colors` being return by `findArrayAttributes` even though `marker.colors` is not in the user data. See https://codepen.io/bpostlethwaite/pen/dJozzo for an example.

This is seemingly inconsistent with standard fullData behaviour.

**Comment by alexcjohnson** [comment]
Possible other cases to worry about:
```
$ ag "dflt: \[\]"
traces/sankey/attributes.js
103:            dflt: [],
161:            dflt: [],
198:            dflt: [],
204:            dflt: [],
209:            dflt: [],

traces/ohlc/attributes.js
67:        dflt: [],
74:        dflt: [],
81:        dflt: [],
88:        dflt: [],

traces/parcoords/attributes.js
125:            dflt: [],

traces/table/attributes.js
72:            dflt: [],
83:            dflt: [],
152:            dflt: [],
163:            dflt: [],

transforms/groupby.js
32:        dflt: [],
```

A lot of these cases the trace wouldn't make sense without the attribute but... then what's the point of filling in an empty array?

**Comment by etpinard** [comment]
Alternatively, maybe we should skip over empty arrays [here](https://github.com/plotly/plotly.js/blob/599dff93f9923e7ef4d6446de734cfb4eb39d663/src/plot_api/plot_schema.js#L194).

**Comment by alexcjohnson** [comment]
> skip over empty arrays

Worries me a bit, empty arrays can, in some cases, be different from non-arrays or missing values.

**Comment by etpinard** [comment]
This appears fixed in `1.34.0` (most likely by the `Plotly.react` push)

![image](https://user-images.githubusercontent.com/6675409/36813098-6246c026-1ca1-11e8-8b24-d1f33188ef49.png)

compared to in `1.33.1`:

![image](https://user-images.githubusercontent.com/6675409/36813118-7d0008a0-1ca1-11e8-8a96-d0e3cc868e7e.png)


Closing.
